File: src/IsbInterop.CSharp/ThreadUtils.cs
using System;
using System.Threading;

namespace IsbInterop
{
  /// <summary>
  /// Утилиты для работы с потоком выполнения.
  /// </summary>
  public static class ThreadUtils
  {
    // Альтернативный вариант.
    ///// <summary>
    ///// Выполняет функцию с таймаутом.
    ///// </summary>
    ///// <param name="function">Функция.</param>
    ///// <param name="timeout">Таймаут.</param>
    ///// <returns>Результат.</returns>
    //public static object Execute(Func<object> function, TimeSpan timeout)
    //{
    //  object result = null;
    //  var thread = new Thread(() =>
    //  {
    //    result = function();
    //  });

    //  thread.Start();
    //  var completed = thread.Join(timeout);
    //  if (!completed)
    //    throw new TimeoutException(string.Format("The operation has timed out after {0}.", timeout));

    //  return result;
    //}

    /// <summary>
    /// Выполняет функцию с таймаутом.
    /// </summary>
    /// <typeparam name="T">Тип результата функции.</typeparam>
    /// <param name="function">Функция.</param>
    /// <param name="timeout">Таймаут.</param>
    /// <returns>Результат.</returns>
    public static T Invoke<T>(Func<T> function, TimeSpan timeout)
    {
      var functionResult = function.BeginInvoke(null, null);
      var waitHandle = functionResult.AsyncWaitHandle;
      if (!waitHandle.WaitOne(timeout))
      {
        // IMPORTANT: Always call EndInvoke to complete your asynchronous call.
        // http://msdn.microsoft.com/en-us/library/2e08f6yc(VS.80).aspx
        // (even though we arn't interested in the result)

        ThreadPool.UnsafeRegisterWaitForSingleObject(waitHandle,
            (state, timedOut) => function.EndInvoke(functionResult),
            null, -1, true);

        throw new IsbInteropTimoutException($"The operation has timed out after {timeout}.");
      }

      return function.EndInvoke(functionResult);
    }
  }
}

File: src/IsbInterop/ThreadUtils.vb
Imports System.Threading

''' <summary>
'''Утилиты для работы с потоком выполнения.
''' </summary>
Public NotInheritable Class ThreadUtils

    ''' <summary>
    ''' Выполняет функцию с таймаутом.
    ''' </summary>
    ''' <typeparam name="T">Тип результата функции.</typeparam>
    ''' <param name="function">Функция.</param>
    ''' <param name="timeout">Таймаут.</param>
    ''' <returns>Результат.</returns>
    Public Shared Function Invoke(Of T)([function] As Func(Of T), timeout As TimeSpan) As T
        Dim functionResult = [function].BeginInvoke(Nothing, Nothing)
        Dim waitHandle = functionResult.AsyncWaitHandle
        If Not waitHandle.WaitOne(timeout) Then
            ' IMPORTANT: Always call EndInvoke to complete your asynchronous call.
            ' http://msdn.microsoft.com/en-us/library/2e08f6yc(VS.80).aspx
            ' (even though we arn't interested in the result)

            ThreadPool.UnsafeRegisterWaitForSingleObject(waitHandle, Function(state, timedOut) [function].EndInvoke(functionResult), Nothing, -1, True)

            Throw New IsbInteropTimeoutException(String.Format("The operation has timed out after {0}.", timeout))
        End If

        Return [function].EndInvoke(functionResult)
    End Function

    ''' <summary>
    ''' Приватный конструктор.
    ''' </summary>
    Private Sub New()
    End Sub
End Class
